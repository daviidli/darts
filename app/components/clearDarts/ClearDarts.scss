.clearDartsPanel {
	height: 100%;
	width: 100vw;
	background-color: rgba(0, 0, 0, 0.6);
	position: absolute;
	visibility: hidden;
	margin: 0 !important;

	&.visible {
		visibility: visible;
	}

	.clearDarts {
		font-weight: 500;
	}
}
