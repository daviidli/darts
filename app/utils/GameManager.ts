/* eslint-disable no-await-in-loop */
import DartBoard from './dartBoard';
import store from '../store/store';
import {
	setCurrentRound,
	setCurrentPlayer,
	initializedRound,
	setWinner,
	setWaiting,
	setMaxDarts,
	setMaxRounds,
	setCurrentDart,
	setTotals
} from '../actions/gameActions';

abstract class GameManager {
	private dartboard: DartBoard;
	private running: boolean;
	private missedDart: boolean;
	private dartsCleared: boolean;

	constructor() {
		const state = store.getState();

		if (!state.serialPort.length) {
			throw new Error('No serial port selected.');
		}

		this.dartboard = new DartBoard(state.serialPort);
		this.running = false;
		this.missedDart = false;
		this.dartsCleared = false;
	}

	protected abstract initialize(): void;
	protected abstract async turn(): Promise<void>;

	protected async run() {
		const {
			maxRounds,
			maxDarts,
			players,
			clearDartsWaitTime
		} = store.getState();

		const roundsArr = this.createStrArray(
			players.length,
			maxRounds,
			maxDarts
		);
		store.dispatch(initializedRound(roundsArr));

		this.running = true;
		for (let r = 0; r < maxRounds && this.running; r++) {
			store.dispatch(setCurrentRound(r));
			for (let p = 0; p < players.length && this.running; p++) {
				store.dispatch(setCurrentPlayer(p));
				await this.turn();
				await this.waitForClearingDarts(clearDartsWaitTime);
			}
		}
	}

	protected async getDart(): Promise<string> {
		let currentDart;
		while (true) {
			try {
				currentDart = await this.dartboard.getResults(1, 500);
			} catch (e) {
				// continue regardless of error
				// todo: show error notification
			}

			if (this.missedDart) {
				this.missedDart = false;
				return 'miss';
			}

			if (currentDart) {
				return currentDart[0];
			}
		}
	}

	public start() {
		this.initialize();
		this.run();
	}

	public stop() {
		this.running = false;
		this.dartboard.close();
		store.dispatch(setMaxDarts(0));
		store.dispatch(setMaxRounds(0));
		store.dispatch(setCurrentPlayer(0));
		store.dispatch(setCurrentDart(0));
		store.dispatch(setCurrentRound(0));
		store.dispatch(initializedRound([]));
		store.dispatch(setTotals([]));
	}

	public setWinner(index: number) {
		this.stop();
		store.dispatch(setWinner(index));
	}

	public miss() {
		this.missedDart = true;
	}

	public getRunning() {
		return this.running;
	}

	public clearedDarts() {
		this.dartsCleared = true;
	}

	private async waitForClearingDarts(timeout = 0) {
		this.dartsCleared = false;
		store.dispatch(setWaiting(true));
		let timeouts = 0;

		while (true) {
			await this.timeout(1000);
			timeouts += 1;
			if (timeouts === timeout) {
				store.dispatch(setWaiting(false));
				return;
			}
			if (this.dartsCleared) {
				this.dartsCleared = false;
				store.dispatch(setWaiting(false));
				return;
			}
		}
	}

	private timeout(ms: number): Promise<void> {
		return new Promise(resolve => setTimeout(resolve, ms));
	}

	private createStrArray(
		iMax: number,
		jMax: number,
		kMax: number
	): string[][][] {
		const initialRounds: string[][][] = [];
		for (let i = 0; i < iMax; i++) {
			const arr = [];
			for (let j = 0; j < jMax; j++) {
				arr.push(new Array(kMax).fill(''));
			}
			initialRounds.push(arr);
		}
		return initialRounds;
	}
}

export default GameManager;
