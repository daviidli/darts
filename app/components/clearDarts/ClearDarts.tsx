import React from 'react';
import { Grid, Typography, Button } from '@material-ui/core';
import clsx from 'clsx';
import GameManager from '../../utils/GameManager';
import styles from './ClearDarts.scss';

export type Props = {
	waiting: boolean;
	gameManager: GameManager;
};

const ClearDarts = (props: Props) => {
	const { waiting, gameManager } = props;

	return (
		<Grid
			container
			direction="column"
			justify="center"
			alignItems="center"
			spacing={8}
			className={clsx(styles.clearDartsPanel, {
				[styles.visible]: waiting
			})}
		>
			<Grid item>
				<Typography variant="h1" className={styles.clearDarts}>
					Clear Darts
				</Typography>
			</Grid>
			<Grid item>
				<Button
					variant="contained"
					color="secondary"
					onClick={() => gameManager.clearedDarts()}
				>
					Darts Cleared
				</Button>
			</Grid>
		</Grid>
	);
};

export default ClearDarts;
