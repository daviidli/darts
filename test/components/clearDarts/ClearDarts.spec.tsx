import { spy } from 'sinon';
import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import { Button } from '@material-ui/core';
import ClearDarts, {
	Props
} from '../../../app/components/clearDarts/ClearDarts';
import GameManager from '../../../app/utils/GameManager';

Enzyme.configure({ adapter: new Adapter() });

const initialProps: Props = {
	waiting: false,
	gameManager: {} as GameManager
};

const setup = (props: Props = initialProps) => {
	const wrapper = shallow(<ClearDarts {...props} />);
	return {
		wrapper,
		button: wrapper.find(Button)
	};
};

describe('ClearDarts component', () => {
	it('should match snapshot', () => {
		const { wrapper } = setup();
		expect(toJson(wrapper)).toMatchSnapshot();
	});

	it('should match snapshot when waiting = true', () => {
		const props = { ...initialProps };
		props.waiting = true;
		const { wrapper } = setup(props);
		expect(toJson(wrapper)).toMatchSnapshot();
	});

	it('should trigger clearDarts when button is clicked', () => {
		const clearedDartsMock = spy();
		const props: Props = {
			waiting: true,
			gameManager: { clearedDarts: clearedDartsMock } as any
		};
		const { button } = setup(props);
		expect(button.text()).toMatch('Darts Cleared');
		expect(clearedDartsMock.notCalled).toBeTruthy();
		button.simulate('click');
		expect(clearedDartsMock.calledOnce).toBeTruthy();
	});
});
