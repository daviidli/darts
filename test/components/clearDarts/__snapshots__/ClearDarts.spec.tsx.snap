// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClearDarts component should match snapshot 1`] = `
<WithStyles(ForwardRef(Grid))
  alignItems="center"
  className="clearDartsPanel"
  container={true}
  direction="column"
  justify="center"
  spacing={8}
>
  <WithStyles(ForwardRef(Grid))
    item={true}
  >
    <WithStyles(ForwardRef(Typography))
      className="clearDarts"
      variant="h1"
    >
      Clear Darts
    </WithStyles(ForwardRef(Typography))>
  </WithStyles(ForwardRef(Grid))>
  <WithStyles(ForwardRef(Grid))
    item={true}
  >
    <WithStyles(ForwardRef(Button))
      color="secondary"
      onClick={[Function]}
      variant="contained"
    >
      Darts Cleared
    </WithStyles(ForwardRef(Button))>
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Grid))>
`;

exports[`ClearDarts component should match snapshot when waiting = true 1`] = `
<WithStyles(ForwardRef(Grid))
  alignItems="center"
  className="clearDartsPanel visible"
  container={true}
  direction="column"
  justify="center"
  spacing={8}
>
  <WithStyles(ForwardRef(Grid))
    item={true}
  >
    <WithStyles(ForwardRef(Typography))
      className="clearDarts"
      variant="h1"
    >
      Clear Darts
    </WithStyles(ForwardRef(Typography))>
  </WithStyles(ForwardRef(Grid))>
  <WithStyles(ForwardRef(Grid))
    item={true}
  >
    <WithStyles(ForwardRef(Button))
      color="secondary"
      onClick={[Function]}
      variant="contained"
    >
      Darts Cleared
    </WithStyles(ForwardRef(Button))>
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Grid))>
`;
